src/Adapter/Profile/Permission/QueryHandler/GetPermissionsForConfigurationHandler.php

```diff
@@ -0,0 +1,304 @@
+        $permissions = ['view', 'add', 'edit', 'delete'];
+
+        $tabPermissionsForProfiles = $this->getTabPermissionsForProfiles($profiles);
+        $modulePermissionsForProfiles = $this->getModulePermissionsForProfiles($profiles);
+
+        $permissionIds = ['view' => 0, 'add' => 1, 'edit' => 2, 'delete' => 3, 'all' => 4];
+        $employeeProfileId = $query->getEmployeeProfileId()->getValue();
+        $canEmployeeEditPermissions = $this->authorizationChecker->isGranted(PageVoter::UPDATE, 'AdminAccess');
+
+        $bulkConfigurationPermissions = $this->getBulkConfigurationForProfiles(
+            $employeeProfileId,
+            $canEmployeeEditPermissions,
+            $tabPermissionsForProfiles,
+            $profiles,
+            $tabs,
+            $permissions
+        );
+
+        return new ConfigurablePermissions(
+            $tabPermissionsForProfiles,
+            $modulePermissionsForProfiles,
+            $profiles,
+            $tabs,
+            $bulkConfigurationPermissions,
+            $permissions,
+            $permissionIds,
+            $employeeProfileId,
+            $canEmployeeEditPermissions
+        );
+    }
+
+    /**
+     * @return int[] IDs of non configurable tabs
+     */
+    private function getNonConfigurableTabs()
+    {
+        return [
+            Tab::getIdFromClassName('AdminLogin'),
+        ];
+    }
+
+    /**
+     * @return array
+     */
+    private function getProfilesForPermissionsConfiguration()
+    {
+        $legacyProfiles = Profile::getProfiles(Context::getContext()->language->id);
+        $profiles = [];
+
+        foreach ($legacyProfiles as $profile) {
+            $isAdministrator = (int) $profile['id_profile'] === _PS_ADMIN_PROFILE_;
+
+            $profiles[] = [
+                'id' => $profile['id_profile'],
+                'name' => $profile['name'],
+                'is_administrator' => $isAdministrator,
+            ];
+        }
+
+        return $profiles;
+    }
+
+    /**
+     * @return array
+     */
+    private function getTabsForPermissionsConfiguration()
+    {
+        $nonConfigurableTabs = $this->getNonConfigurableTabs();
+        $legacyTabs = Tab::getTabs(Context::getContext()->language->id);
+        $tabs = [];
+
+        foreach ($legacyTabs as $tab) {
+            // Don't allow permissions for unnamed tabs (ie. AdminLogin)
+            if (empty($tab['name'])) {
+                continue;
+            }
+
+            if (in_array($tab['id_tab'], $nonConfigurableTabs)) {
+                continue;
+            }
+
+            $tabs[] = [
+                'id' => $tab['id_tab'],
+                'id_parent' => $tab['id_parent'],
+                'name' => $tab['name'],
+            ];
+        }
+
+        return $this->buildTabsTree($tabs);
+    }
+
+    /**
+     * @param array $tabs
+     * @param int $parentId
+     *
+     * @return array
+     */
+    private function buildTabsTree(array &$tabs, $parentId = 0)
```

Comment:
If someones screws up his database, he could have a recurvise tab structure that would make this function never end. In order to prevent this, can you add a MAX_NESTED_TAB_TREE_CALLS check ? At each `buildTabsTree()` call you pass a incremented counter, and when it reaches MAX_NESTED_TAB_TREE_CALLS it throws a function.

Reply:
I'm not sure if I understand your idea. :thinking: How do we know value of `MAX_NESTED_TAB_TREE_CALLS`? And I think we should not try to fix what's broken (database in this case), but let user see error instead.

Reply:
```
private function buildTabsTree(array &$tabs, $parentId = 0, $iterationNumber)
    {
        if ($iterationNumber > self::MAX_NESTED_TAB_TREE_CALLS) {
            throw new \RuntimeException('Maximum nested calls reached buildTabsTree');
        }

        $children = [];
        foreach ($tabs as &$tab) {
            $id = $tab['id'];
            if ((int) $tab['id_parent'] === (int) $parentId) {
                $children[$id] = $tab;
                $children[$id]['children'] = $this->buildTabsTree($tabs, $id, $iterationNumber+1);
            }
        }
        return $children;
    }
```
It needs to be initialized in 1st call. The idea is that everytime this functionn calls itself, we increase `$iterationNumber` to monitor the number of nested calls and stop it when it's too high.

Without this, this php file can trigger an out-of-memory error very hard to debug 😉

Reply:
Except you can't have a non optional argument at the end, besides I think it will be more convenient to have a signature like `private function buildTabsTree(array &$tabs, $parentId = 0, $iterationNumber = 0)`
So you don't need to init the counter on the first call

Reply:
I've addressed this comment & added description in commit, should we document this decision in code as well?